### PR TITLE
Raw mapping merge fix for properties field

### LINKED ISSUE #108866: Failure to merge mappings of field named `properties`
### Elasticsearch Version

8.14.0

### Installed Plugins

_No response_

### Java Version

_bundled_

### OS Version

Darwin

### Problem Description

The new merge algorithm of raw field mappings that was merged through #97317 introduced a new failure when merging mappings of fields named `properties`. See details in the [comment](https://github.com/elastic/elasticsearch/pull/97317#issuecomment-2112857574) posted to the related PR.

### Steps to Reproduce

Merging:
```
  "template": {
    "mappings": {
      "properties": {
        "properties": {
          "properties": {
            "child1": {
              "type": "text"
            }
          }
        }
      }
    }
  }
```
with:
```
  "template": {
    "mappings": {
      "properties": {
        "properties": {
          "properties": {
            "child2": {
              "type": "long"
            }
          }
        }
      }
    }
  }
```
should yield:
```
  "template": {
    "mappings": {
      "properties": {
        "properties": {
          "properties": {
            "child1": {
              "type": "text"
            },
            "child2": {
              "type": "long"
            }
          }
        }
      }
    }
  }
```
but instead yields:
```
  "template": {
    "mappings": {
      "properties": {
        "properties": {
          "properties": {
            "child2": {
              "type": "long"
            }
          }
        }
      }
    }
  }
```

### Logs (if relevant)

_No response_

### PR DESCRIPTION
**Pull request by eyalkoren**
Fixes #108866

### CODE REVIEW

**Review by felixbarny — APPROVED** [review]
LGTM

**Comment by eyalkoren** [comment]
cc @chiarch84

**Review by cbuescher — APPROVED** [review]
I took a deeper look at the custom merge logic in RawFieldMappingMerge and agree that passing on "null" in the recursion when the parent is a "properties" field should not affect any further merging logic down the line. 
LGTM for that reason.

**Comment by eyalkoren** [comment]
> I took a deeper look at the custom merge logic in RawFieldMappingMerge and agree that passing on "null" in the recursion when the parent is a "properties" field should not affect any further merging logic down the line.

Thanks for verifying 🙏 
Maybe another way to put it is that `XContentHelper.merge()` provides tree-merge functionality  and the optional `parent` argument can be used to add context for the merge **if necessary**. When merging mapping subtrees of a field, we only need the plain merge functionality, so this additional context is not required. 
Therefore, there is already a merge API to do just that, so I just merged a tiny change to use it instead. I think it makes it a bit cleaner. I am asking for a re-review just for sanity.

File: server/src/main/java/org/elasticsearch/index/mapper/MapperService.java [hunk 1]
```diff
@@ -480,8 +480,11 @@ public Object merge(String parent, String key, Object oldValue, Object newValue)
                         if (baseMap.containsKey("subobjects")) {
                             mergedMappings.put("subobjects", baseMap.get("subobjects"));
                         }
-                        // recursively merge these two field mappings
-                        XContentHelper.merge(key, mergedMappings, mapToMerge, INSTANCE);
+                        // Recursively merge these two field mappings.
+                        // Since "key" is an arbitrary field name, for which we only need plain mapping subtrees merge, no need to pass it
+                        // to the recursion as it shouldn't affect the merge logic. Specifically, passing "null" as parent avoids merge
```
Comment (felixbarny) [inline server/src/main/java/org/elasticsearch/index/mapper/MapperService.java:485]: The comment still talks about passing "null" as parent

**Review by felixbarny — APPROVED** [review]
Just a nit on a comment, otherwise LGTM

**Comment by felixbarny** [comment]
Tip: you can apply the `auto-backport-and-merge` label so that the PR automatically gets back ported to 8.14.1 after it is merged.

**Comment by eyalkoren** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.14|[](https://github.com/elastic/elasticsearch/pull/109147)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

**Pull request merged**